hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenBurnHandler.java

```diff
@@ -56,8 +80,98 @@ public void preHandle(@NonNull final PreHandleContext context) throws PreCheckEx
+        if (fungibleCount < 1 && !serialNums.isEmpty()) {
+            for (final var serialNumber : op.serialNumbers()) {
+                validateTruePreCheck(serialNumber > 0, INVALID_NFT_ID);
+            }
+        }
+    }
+
     @Override
     public void handle(@NonNull final HandleContext context) throws HandleException {
-        throw new UnsupportedOperationException("Not implemented");
+        requireNonNull(context);
+        final var accountStore = context.writableStore(WritableAccountStore.class);
+        final var tokenStore = context.writableStore(WritableTokenStore.class);
+        final var tokenRelStore = context.writableStore(WritableTokenRelationStore.class);
+        final var nftStore = context.writableStore(WritableNftStore.class);
+        final var txn = context.body();
+        final var op = txn.tokenBurnOrThrow();
+        final var tokenId = op.token();
+        final var fungibleBurnCount = op.amount();
+        // Wrapping the serial nums this way de-duplicates the serial nums:
+        final var nftSerialNums = new ArrayList<>(new LinkedHashSet<>(op.serialNumbers()));
+        final var validated = validateSemantics(tokenId, fungibleBurnCount, nftSerialNums, tokenStore, tokenRelStore);
+        final var token = validated.token();
+
+        if (token.tokenType() == TokenType.FUNGIBLE_COMMON) {
+            changeSupply(
+                    validated.token(),
+                    validated.tokenTreasuryRel(),
+                    -fungibleBurnCount,
+                    INVALID_TOKEN_BURN_AMOUNT,
+                    accountStore,
+                    tokenStore,
+                    tokenRelStore);
+        } else {
+            // Load and validate the nfts
+            for (final Long nftSerial : nftSerialNums) {
+                final var nft = nftStore.get(asUniqueTokenId(tokenId, nftSerial));
+                validateTrue(nft != null, INVALID_NFT_ID);
+
+                final var nftOwner = nft.ownerNumber();
+                validateTrue(treasuryOwnsNft(nftOwner), TREASURY_MUST_OWN_BURNED_NFT);
+            }
+
+            // Update counts for accounts and token rels
+            changeSupply(
+                    token,
+                    validated.tokenTreasuryRel(),
+                    -nftSerialNums.size(),
+                    FAIL_INVALID,
+                    accountStore,
+                    tokenStore,
+                    tokenRelStore);
+
+            // Remove the nft objects
+            nftSerialNums.forEach(serialNum -> nftStore.remove(asUniqueTokenId(tokenId, serialNum)));
+        }
     }
+
+    private ValidationResult validateSemantics(
+            @NonNull final TokenID tokenId,
+            final long fungibleBurnCount,
+            @NonNull final List<Long> nftSerialNums,
+            final ReadableTokenStore tokenStore,
+            final ReadableTokenRelationStore tokenRelStore) {
+        validateTrue(fungibleBurnCount > -1, INVALID_TOKEN_BURN_AMOUNT);
```

Comment:
nit : Probably more readable with `>=0` 😃 or may be it is just me

Reply:
Personally I don't think it's very readable either way, but I'm fine with `>=0`

Reply:
Done